server/config/initializers/ckb/utils.rb

```diff
@@ -0,0 +1,98 @@
+
+module CKB
+  module Utils
+    def self.calculate_cell_min_capacity(output)
+      output = output.deep_symbolize_keys
+      capacity = 8 + output[:data].bytesize + calculate_script_capacity(output[:lock])
+      if type = output[:type]
+        capacity += calculate_script_capacity(type)
+      end
+      capacity
+    end
+
+    def self.calculate_script_capacity(script)
+      capacity = 1 + (script[:args] || []).map { |arg| arg.bytesize }.reduce(0, &:+)
+      if script[:binary_hash]
+        capacity += CKB::Utils.hex_to_bin(script[:binary_hash]).bytesize
+      end
+      capacity
+    end
+
+    def self.block_cell_consumed(commit_transactions)
+      commit_transactions.inject(0) do |memo, commit_transaction|
+        memo + commit_transaction["outputs"].reduce(0) { |inside_memo, output| inside_memo + CKB::Utils.calculate_cell_min_capacity(output) }
+      end
+    end
+
+    def self.total_cell_capacity(commit_transactions)
+      commit_transactions.inject(0) do |memo, commit_transaction|
+        memo + commit_transaction["outputs"].reduce(0) { |inside_memo, output| inside_memo + output["capacity"] }
+      end
+    end
+
+    def self.miner_hash(cellbase)
+      CKB::Utils.json_script_to_type_hash(cellbase["outputs"].first["lock"].symbolize_keys)
+    end
+
+    def self.miner_reward(cellbase)
+      cellbase["outputs"].first["capacity"]
+    end
+
+    def self.transaction_fee(transaction)
+      output_capacities = transaction["outputs"].map { |output| output["capacity"] }.reduce(0, &:+)
+      input_capacities = transaction["inputs"].map { |input| CKB::Utils.cell_input_capacity(input) }.reduce(0, &:+)
+      output_capacities - (input_capacities.zero? ? output_capacities : input_capacities)
+    end
+
+    def self.total_transaction_fee(transactions)
+      transactions.inject(0) { |memo, transaction| memo + CKB::Utils.transaction_fee(transaction) }
+    end
+
+    def self.get_unspent_cells(lock_hash)
+      to = CkbSync::Api.get_tip_block_number
+      results = []
+      current_from = 1
+      while current_from <= to
+        current_to = [current_from + 100, to].min
+        cells = CkbSync::Api.get_cells_by_lock_hash(lock_hash, current_from, current_to)
+        results.concat(cells)
+        current_from = current_to + 1
+      end
+      results
+    end
+
+    #TODO 可以改成通过本地的 cell 来计算。
```

Comment:
Annotation should be english as long as possible.

Reply:
ok

Reply:
Update this.